### ISSUE TITLE
InstanceNorm FP16 variance overflow collapses finite feature maps

### ISSUE THREAD

**Issue by ternaus** [ml-explore/mlx#4228]
### Problem

`mlx.nn.InstanceNorm` computes `mean` and `var` in the input dtype. For FP16 feature maps, the variance can overflow even when every input value is finite. The normalization then collapses a non-constant channel to zero. In a trained 3D segmentation model, larger feature maps also produced non-finite outputs.

### Minimal reproducer

Tested with MLX 0.32.0, Apple M4, macOS 26.4.1:

```python
import mlx.core as mx
import mlx.nn as nn
import numpy as np

checkerboard = np.indices((4, 4, 4)).sum(0) % 2
x = np.stack(
    (
        np.where(checkerboard, -512, 512),
        np.where(checkerboard, -256, 256),
    ),
    axis=-1,
)[None].astype(np.float16)
x = mx.array(x)

direct = nn.InstanceNorm(2)(x)
fp32_stats = nn.InstanceNorm(2)(x.astype(mx.float32)).astype(mx.float16)
mx.eval(direct, fp32_stats)

print(np.asarray(mx.var(x, axis=(1, 2, 3))))
print(np.unique(np.asarray(direct)))
print(np.asarray(fp32_stats).min(), np.asarray(fp32_stats).max())
```

Output:

```text
[[inf inf]]
[-0.]
-1.0 1.0
```

### Expected behavior

For FP16 and BF16 inputs, `InstanceNorm` should calculate reduction statistics in FP32, normalize in FP32, then return the input dtype. This matches the mixed-precision policy normally used for normalization layers and prevents a finite, non-constant feature map from collapsing because the variance reduction overflowed.

### Real model impact

The issue appeared while porting a 29.7M-parameter 3D nnU-Net-v1 inference graph from PyTorch MPS to MLX. Full FP16 produced non-finite logits with trained weights. Casting only the `InstanceNorm` input to FP32 for statistics and casting its output back to the feature dtype removed the non-finite values.

On a complete `128 x 96 x 96` FramePack decode-transfer-inference-host route, this mixed MLX path still processed 6.47 patches/s versus 3.48 patches/s for PyTorch MPS FP16 autocast on the same M4.

I can submit a focused PR with an FP16 regression test and before/after benchmark if this dtype policy is acceptable.
